### PR TITLE
feat(I8N) - Add internationalization

### LINKED ISSUE #634: Add Internationalization
Adds Internationalization features such as RTL support

### PR DESCRIPTION
**Pull request by mpodwysocki**
Adds Internationalization features to detect whether right to left.  Closes #634

### CODE REVIEW

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
+﻿using ReactNative.Bridge;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace ReactNative.Modules.Internationalization
+{
+    /// <summary>
+    /// A module that allows JS to get/set right to left preferences.
+    /// </summary>
+    class InternationalizationModule : NativeModuleBase
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs]: >InternationalizationModule [](start = 10, length = 26)

I sort of don't like that the name is so much different from the JS module name.  Is there a precedent for that in other modules?

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        private const string ModuleName = "I18nManager";
+        private const string IsRtl = "isRTL";
+        private const string LocalIdentifier = "localeIdentifier";
+
+
+        internal InternationalizationModule()
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs]: constructor needed?

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+namespace ReactNative.Modules.Internationalization
+{
+    /// <summary>
+    /// A module that allows JS to get/set right to left preferences.
+    /// </summary>
+    class InternationalizationModule : NativeModuleBase
+    {
+        private const string ModuleName = "I18nManager";
+        private const string IsRtl = "isRTL";
+        private const string LocalIdentifier = "localeIdentifier";
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs]: extra newline after this field.

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        /// </summary>
+        public override string Name
+        {
+            get
+            {
+                return ModuleName;
+            }
+        }
+
+        public override IReadOnlyDictionary<string, object> Constants
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs]: Inconsistent application of doc comments.

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs (hunk 1 above)
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs]: And In-ter-na-tion-al-i-za-tion has more syllables than I-eight-teen-N, so it's harder to read :-)

---
In reply to: [76422044](https://github.com/ReactWindows/react-native-windows/pull/640#discussion_r76422044) [](ancestors = 76422044)

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    /// </summary>
+    static class InternationalizationUtil
+    {
+        private const string AllowRTL = "RCTI18nUtil_allowRTL";
+        private const string ForceRTL = "RCTI18nUtil_forceRTL";
+
+        static InternationalizationUtil()
+        {
+            if (ApplicationData.Current.LocalSettings.Values[AllowRTL] == null)
+                ApplicationData.Current.LocalSettings.Values[AllowRTL] = false;
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs]: Can we just leave these blank and do null coalescing when reading out of LocalSettings below?

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        /// <summary>
+        /// Check if the system is using Right to Left. This only happens when the app:
+        /// - is forcing RTL layout regardless of language
+        /// - allows RTL layout when using RTL locale.
+        /// </summary>
+        public static bool IsRightToLeft
+        {
+            get
+            {
+                if (IsRightToLeftForced) return true;
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs]: return IsRightToLeftForced || IsRightToLeftAllowed && IsDeviceRightToLeft (&& has higher operator precedence)

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        /// <summary>
+        /// Gets or sets whether right to left is allowed.  Should be used very early during startup before the bridge is initialized.
+        /// </summary>
+        public static bool IsRightToLeftAllowed
+        {
+            get
+            {
+                return (bool)ApplicationData.Current.LocalSettings.Values[AllowRTL];
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs]: From above, e.g., return (bool?)ApplicationData.Current.LocalSettings.Values[AllowRTL] ?? false;

File: ReactWindows/ReactNative/UIManager/UIImplementation.cs [hunk 8]
```diff
@@ -659,6 +660,9 @@ private void HandleCreateView(ReactShadowNode cssNode, int rootViewTag, ReactSty
         private ReactShadowNode CreateRootShadowNode()
         {
             var rootCssNode = new ReactShadowNode();
+            if (InternationalizationUtil.IsRightToLeft)
```
Comment (rozele) [inline ReactWindows/ReactNative/UIManager/UIImplementation.cs]: Use { }

File: ReactWindows/ReactNative/UIManager/UIImplementation.cs [hunk 9]
```diff
@@ -659,6 +660,9 @@ private void HandleCreateView(ReactShadowNode cssNode, int rootViewTag, ReactSty
...
+                rootCssNode.Direction = Facebook.CSSLayout.CSSDirection.RTL;
```
Comment (rozele) [inline ReactWindows/ReactNative/UIManager/UIImplementation.cs]: Have you confirmed if the CSSDirection is inherited? We use different binaries than are used for iOS and Android.

Comment (rozele) [inline ReactWindows/ReactNative/UIManager/UIImplementation.cs]: >Facebook.CSSLayout [](start = 40, length = 18)

using Facebook.CSSLayout;

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs (hunk 1 above)
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs]: @rozele renaming to `I18NUtil` and `I18NModule` in the `I18N` namespace

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs (hunk 2 above)
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationModule.cs]: @rozele might want to revisit the `ClipboardModule` which has the same issue

File: ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs (hunk 5 above)
Comment (mpodwysocki) [inline ReactWindows/ReactNative/Modules/Internationalization/InternationalizationUtil.cs]: @rozele depends where we want to take a hit, once per app install or once per check.

File: ReactWindows/ReactNative/Modules/I18N/I18NModule.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+namespace ReactNative.Modules.I18N
+{
+    /// <summary>
+    /// A module that allows JS to get/set right to left preferences.
+    /// </summary>
+    class I18NModule : NativeModuleBase
+    {
+        private const string ModuleName = "I18nManager";
+        private const string IsRtl = "isRTL";
+        private const string LocalIdentifier = "localeIdentifier";
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/I18N/I18NModule.cs]: Extra newline here.

File: ReactWindows/ReactNative/Modules/I18N/I18NUtil.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace ReactNative.Modules.I18N
+{
+    /// <summary>
+    /// Class used for right to left preferences.
+    /// </summary>
+    static class I18NUtil
+    {
+        private const string AllowRTL = "RCTI18nUtil_allowRTL";
+        private const string ForceRTL = "RCTI18nUtil_forceRTL";
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/I18N/I18NUtil.cs]: Extra newline.

**Comment by rozele** [comment]
LGTM :shipit:

File: ReactWindows/ReactNative/Modules/I18N/I18NUtil.cs [hunk 12]
```diff
@@ -8,6 +8,7 @@ namespace ReactNative.Modules.I18N
     /// </summary>
     static class I18NUtil
     {
+
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/I18N/I18NUtil.cs]: This isn't what I meant.

**Pull request merged**